content/learning-paths/smartphones-and-mobile/ray_tracing/rt04_acceleration_structure.md

```diff
@@ -0,0 +1,191 @@
+    }
+    as_geom_info.scratchData += scratch_buffer.get_offset();        // ensure it is aligned
+
+    as_geom_info.dstAccelerationStructure = as_handle;
+
+    std::vector<VkAccelerationStructureBuildRangeInfoKHR> build_range_infos{};
+    build_range_infos.reserve(geometry_primitive_counts.size());
+    for (uint32_t geom_primitive_count : geometry_primitive_counts)
+    {
+        VkAccelerationStructureBuildRangeInfoKHR geom_range_info{};
+        geom_range_info.primitiveCount = geom_primitive_count;
+        build_range_infos.emplace_back(geom_range_info);
+    }
+
+    if (build_on_host)
+    {
+        vkBuildAccelerationStructuresKHR(device, VK_NULL_HANDLE, 1, &as_geom_info, &build_range_infos.data());
+    }
+    else
+    {
+        vkCmdBuildAccelerationStructuresKHR(command_buffer, 1, &as_geom_info, &build_range_infos.data());
+    }
+}
+```
+
+Finally, the quality of your geometry can have a huge impact when building your acceleration structure. Try to reduce the number of vertices as much as possible. Grouping objects close together into the same BLAS can also have a positive impact. At the same time, try to reduce overlapping across BLASes and reduce empty space inside a BLAS as much as possible. Empty space will increase the size of a BLAS, which might cause extra AABB hits. Similarly, if BLASes overlap, we might need to evaluate triangles on multiples BLASes.
+
+{{< tabpane >}}
+  {{< tab header="Example of a scene with a bad acceleration structure" img_src="/learning-paths/smartphones-and-mobile/ray_tracing/images/city_scene_as_overlapping.png" title="Example of a scene with a bad acceleration structure: Significant overlap and empty space. This will result in unnecessary triangle intersections." >}}{{< /tab >}}
+  {{< tab header="Example of a scene with a good acceleration structure" img_src="/learning-paths/smartphones-and-mobile/ray_tracing/images/city_scene_as_separate.png" title="Example of a scene with a good acceleration structure: Minimum overlap and empty space." >}}{{< /tab >}}
+{{< /tabpane >}}
+
+When tracing a scene, it is common to have opaque and non-opaque objects. Non opaque objects will invoke the `Any-Hit` shader and require extra evaluation in ray query. Vulkan allows to use the flags `VK_GEOMETRY_INSTANCE_FORCE_OPAQUE_BIT_KHR` and `VK_GEOMETRY_OPAQUE_BIT_KHR` to mark specific instances and geometries as opaque. We recommend using the appropriate flags to mark relevant geometry as opaque.
+
+Vulkan also offers shader flags, they will affect all objects, but they offer better performance. You can use `gl_RayFlagsCullNoOpaqueEXT` to cull no opaque geometries or `gl_RayFlagsOpaqueEXT`, to mark all objects as opaque. To enable compiler optimizations you should also use `gl_RayFlagsSkipAABBEXT` and ensure the flags can be resolved statically, if possible, prefer `gl_RayFlagsCullNoOpaqueEXT` it might improve shader performance.
```

Comment:
"to cull no opaque" -> "to cull non-opaque"
Last sentence... split in 2?
"To enable compiler optimizations you should also use `gl_RayFlagsSkipAABBEXT` and ensure the flags can be resolved statically. If possible, use `gl_RayFlagsCullNoOpaqueEXT` as it might improve shader performance."